### PR TITLE
TCP Transport will no longer hang when its port is already in use

### LINKED ISSUE #1388: Gossip server goes to infinite loop on address already in use
`serverMainLoop` should signal to `NewDirectTransport` that the port is taken, so the loop doesn't run forever.

https://circleci.com/gh/orbs-network/orbs-network-go/20341?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

### PR DESCRIPTION
**Pull request by electricmonk**
fixes #1388

### CODE REVIEW

File: services/gossip/adapter/tcp/direct_transport_test.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ func TestDirectTransport_HandlesStartupWithEmptyPeerList(t *testing.T) {
 		defer transport.GracefulShutdown(ctx)
 
 		require.True(t, test.Eventually(test.EVENTUALLY_ADAPTER_TIMEOUT, func() bool {
-			return transport.IsServerListening()
+			return transport.server.IsListening()
```
Comment (amir-arad) [inline services/gossip/adapter/tcp/direct_transport_test.go]: why not use  `IsServerListening()` in tests?

File: services/gossip/adapter/tcp/server.go [hunk 2]
```diff
@@ -12,56 +12,91 @@ import (
...
+type transportServer struct {
+	sync.RWMutex
+	listener                          adapter.TransportListener
+	listening                         bool
+	port                              int
+	logger                            log.Logger
+	incomingConnectionAcceptSuccesses *metric.Gauge
+	incomingConnectionAcceptErrors    *metric.Gauge
+	incomingConnectionTransportErrors *metric.Gauge
+	activeIncomingConnections         *metric.Gauge
```
Comment (amir-arad) [inline services/gossip/adapter/tcp/server.go]: separate metrics from core functionality

File: services/gossip/adapter/tcp/server.go [hunk 3]
```diff
@@ -12,56 +12,91 @@ import (
...
 )
 
-func (t *DirectTransport) serverListenForIncomingConnections(ctx context.Context, listenPort uint16) (net.Listener, error) {
+type serverConfig interface {
+	GossipListenPort() uint16
+	GossipNetworkTimeout() time.Duration
+}
+
+type transportServer struct {
+	sync.RWMutex
```
Comment (amir-arad) [inline services/gossip/adapter/tcp/server.go]: ideally the mutex (along with `lock()`, `unlock()`) should be a private implementation detail of the server.

File: services/gossip/adapter/tcp/server.go [hunk 4]
```diff
@@ -12,56 +12,91 @@ import (
...
+		incomingConnectionAcceptSuccesses: registry.NewGauge("Gossip.IncomingConnection.ListeningOnTCPPortSuccess.Count"),
+		incomingConnectionAcceptErrors:    registry.NewGauge("Gossip.IncomingConnection.ListeningOnTCPPortErrors.Count"),
+		incomingConnectionTransportErrors: registry.NewGauge("Gossip.IncomingConnection.TransportErrors.Count"),
+		activeIncomingConnections:         registry.NewGauge("Gossip.IncomingConnection.Active.Count"),
+	}
+}
+
+func (t *transportServer) getPort() int {
+	t.Lock()
+	defer t.Unlock()
```
Comment (amir-arad) [inline services/gossip/adapter/tcp/server.go]: shouldn't that be `rLock()` and `rUnlock()`?

File: services/gossip/adapter/tcp/server.go (hunk 3 above)
Comment (electricmonk) [reply services/gossip/adapter/tcp/server.go]: It is. Embedding a mutex is the idiomatic practice.

**Pull request merged**